File: Laborator/2020-2021/lab01 - Divide et Impera/skel/cpp/task-4/main.cpp
#include <iostream>
#include <fstream>
using namespace std;

class Task {
public:
    void solve() {
        read_input();
        print_output(fast_pow(base, exponent, mod));
    }

private:
    int base, exponent, mod;

    void read_input() {
        ifstream fin("in");
        fin >> base >> exponent >> mod;
        fin.close();
    }

    int fast_pow(int base, int exponent, int mod) {
        // TODO: Calculati (base ^ exponent) % mod in O(log exponent)
        int ans = 1;
        while(exponent != 0) {
            if (exponent % 2 == 1) {
                ans = (1LL * (ans % mod) * (base % mod)) % mod;
            }
            base = (1LL * (base % mod) * (base % mod)) % mod;
            exponent /= 2;
        }
        return ans;
    }

    void print_output(int result) {
        ofstream fout("out");
        fout << result;
        fout.close();
    }
};

// Please always keep this simple main function!
int main() {
    // Allocate a Task object on heap in order to be able to
    // declare huge static-allocated data structures inside the class.
    Task *task = new Task();
    task->solve();
    delete task;
    return 0;
}

File: Laborator/2020-2021/lab01 - Divide et Impera/skel/java/task-4/src/Main.java
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner;

public class Main {
    static class Task {
        public final static String INPUT_FILE = "in";
        public final static String OUTPUT_FILE = "out";

        int base;
        int exponent;
        int mod;

        public void solve() {
            readInput();
            writeOutput(getResult());
        }

        private void readInput() {
            try {
                Scanner sc = new Scanner(new File(INPUT_FILE));
                base = sc.nextInt();
                exponent = sc.nextInt();
                mod = sc.nextInt();
                sc.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private void writeOutput(int result) {
            try {
                PrintWriter pw = new PrintWriter(new File(OUTPUT_FILE));
                pw.printf("%d\n", result);
                pw.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private int fastPow(long base, int exponent, int mod) {
            // TODO: Calculati (base^exponent) modulo mod in O(log exponent).
            long ans = 1;
            while(exponent != 0) {
                if (exponent % 2 == 1) {
                    ans = (1L * (ans % mod) * (base % mod)) % mod;
                }
                base = (1L * (base % mod) * (base % mod)) % mod;
                exponent /= 2;
            }
            return (int)ans;
        }

        private int getResult() {
            return fastPow(base, exponent, mod);
        }
    }

    public static void main(String[] args) {
        new Task().solve();
    }
}
